### PR TITLE
mutation: Improve log print of mutations

### LINKED ISSUE #826: Improve printout of mutations
Currently it's not very human-readable. Cells and keys are displayed as serialized blobs, column names are not showed:

```
TRACE [shard 0] database - apply {ks.users key {key: pk{00000054}, token:-2255749252216079110} data {mutation_parti
tion: {tombstone: none} () static {row: } clustered {rows_entry: ckp{} {deletable_row: {row_marker 1453205401231155
 0 0} {tombstone: none} {row: {column: 0 01000529aebfb9df336b3834}, {column: 1 01000529aebfb9df336b3834}}}}}}
```

Something like this would be nicer:

```
TRACE [shard 0] database - apply {table: 'ks.users', key: {'pk': 'k1', token: -2255749252216079110}, partition: {
    rows: [
        {created_at: 1453205401231155, cells: {
            'firstName': {timestamp: 1453205401231155, value: 'John'},
            'lastName': {timestamp: 1453205401231155, value: 'Smith'}
        }}
    ]
}}

```

### PR DESCRIPTION
**Pull request by avelanarius**
Changes format of `mutation`, `mutation_partition`
log messages to more human-readable. Fixes #826.

##### Example 1
Before:
```
TRACE 2020-09-01 16:30:12,411 [shard 8] database - apply {cycling.t2 {key: pk{000400000003}, token:9010454139840013625} {mutation_partition: 
 static: cont=1 {row: },
 clustered: {
    {rows_entry: cont=true dummy=false {position: clustered,5, 0} {deletable_row: {row_marker: 1598970612410213 0 0} {row: 
      {v, { atomic_cell{7,ts=1598970612410213,expiry=-1,ttl=0} }}}}}}}}
```
After:
```
TRACE 2020-09-04 13:17:47,274 [shard 8] database - apply {table: 'cycling.t', key: {'pk': 3, token: 9010454139840013625}, mutation_partition: {
  rows: [
    {
      cont: true,
      dummy: false,
      marker: {row_marker: 1599218267273611 0 0},
      position: {
        bound_weight: 0,
        'ck': 5,
      },
      'v': { atomic_cell{7,ts=1599218267273611,expiry=-1,ttl=0} },
    },
  ]
}
}
```

##### Example 2
Before (part):
```
columns={system_schema.columns {key: pk{001273797374656d5f6469737472696275746564}, token:1877167950303559708} {mutation_partition: 
 static: cont=1 {row: },
 clustered: {
    {rows_entry: cont=true dummy=false {position: clustered,cdc_generation_descriptions:description, 0} {deletable_row: {row: 
      {clustering_order, { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {column_name_bytes, { atomic_cell{6465736372697074696f6e,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {kind, { atomic_cell{regular,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {position, { atomic_cell{-1,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {type, { atomic_cell{frozen<list<frozen<tuple<bigint, frozen<list<blob>>, tinyint>>>>,ts=-9223372036854775807,expiry=-1,ttl=0} }}}}},
    {rows_entry: cont=true dummy=false {position: clustered,cdc_generation_descriptions:expired, 0} {deletable_row: {row: 
      {clustering_order, { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {column_name_bytes, { atomic_cell{65787069726564,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {kind, { atomic_cell{regular,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {position, { atomic_cell{-1,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {type, { atomic_cell{timestamp,ts=-9223372036854775807,expiry=-1,ttl=0} }}}}},
    {rows_entry: cont=true dummy=false {position: clustered,cdc_generation_descriptions:time, 0} {deletable_row: {row: 
      {clustering_order, { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {column_name_bytes, { atomic_cell{74696d65,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {kind, { atomic_cell{partition_key,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {position, { atomic_cell{0,ts=-9223372036854775807,expiry=-1,ttl=0} }},
      {type, { atomic_cell{timestamp,ts=-9223372036854775807,expiry=-1,ttl=0} }}}}}}}},
},
```
After (part):
```
 columns={table: 'system_schema.columns', key: {'keyspace_name': system_distributed, token: 1877167950303559708}, mutation_partition: {
  rows: [
    {
      cont: true,
      dummy: false,
      position: {
        bound_weight: 0,
        'table_name': cdc_generation_descriptions,
        'column_name': description,
      },
      'clustering_order': { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'column_name_bytes': { atomic_cell{6465736372697074696f6e,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'kind': { atomic_cell{regular,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'position': { atomic_cell{-1,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'type': { atomic_cell{frozen<list<frozen<tuple<bigint, frozen<list<blob>>, tinyint>>>>,ts=-9223372036854775807,expiry=-1,ttl=0} },
    },
    {
      cont: true,
      dummy: false,
      position: {
        bound_weight: 0,
        'table_name': cdc_generation_descriptions,
        'column_name': expired,
      },
      'clustering_order': { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'column_name_bytes': { atomic_cell{65787069726564,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'kind': { atomic_cell{regular,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'position': { atomic_cell{-1,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'type': { atomic_cell{timestamp,ts=-9223372036854775807,expiry=-1,ttl=0} },
    },
    {
      cont: true,
      dummy: false,
      position: {
        bound_weight: 0,
        'table_name': cdc_generation_descriptions,
        'column_name': time,
      },
      'clustering_order': { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'column_name_bytes': { atomic_cell{74696d65,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'kind': { atomic_cell{partition_key,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'position': { atomic_cell{0,ts=-9223372036854775807,expiry=-1,ttl=0} },
      'type': { atomic_cell{timestamp,ts=-9223372036854775807,expiry=-1,ttl=0} },
    },
  ]
}
}
```

### CODE REVIEW

File: mutation_partition.cc [hunk 1]
```diff
@@ -1028,19 +1028,65 @@ operator<<(std::ostream& os, const rows_entry::printer& p) {
...
+            os << indent << indent << indent <<  "'" << column_def.name_as_text() 
+               << "': " << atomic_cell_or_collection::printer(column_def, cell) << ",\n";
+        });
+
+        os << indent << indent << "},\n";
+    }
+
+    os << indent << "]";
+
+    if (!mp.static_row().empty()) {
```
Comment (haaawk) [inline mutation_partition.cc:1077]: Did you swap clustering rows and static row order?

File: mutation_partition.cc [hunk 2]
```diff
@@ -1028,19 +1028,65 @@ operator<<(std::ostream& os, const rows_entry::printer& p) {
...
+            auto ck = *pip.get_clustering_key_prefix();
+            auto type_iterator = ck.get_compound_type(p._schema)->types().begin();
+            auto column_iterator = p._schema.clustering_key_columns().begin();
+
+            for (auto&& e : ck.components(p._schema)) {
+                os << indent << indent << indent << "'" << column_iterator->name_as_text() 
+                   << "': " << (*type_iterator)->to_string(to_bytes(e)) << ",\n";
+                ++type_iterator;
+                ++column_iterator;
+            }
```
Comment (haaawk) [inline mutation_partition.cc:1063]: I think I've seen this code before above, no? Maybe encapsulate the logic into a function to avoid code duplication?

**Review by haaawk — APPROVED** [review]
LGTM. Only the order of static/clustered rows seems not right.

**Comment by denesb** [comment]
>  columns={table: 'system_schema.columns', key: {'keyspace_name': system_distributed, token: 1877167950303559708}, mutation_partition: {
>     rows: [
>         {
>             'table_name': cdc_generation_descriptions,
>             'column_name': description,
>             'clustering_order': { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} }


This is almost JSON, maybe it would be worth it to make it real JSON so we could load it in a parser and slice it process it, etc. We'd have to tweak the atomic_cell printout as well for that.

>             'column_name_bytes': { atomic_cell{6465736372697074696f6e,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'kind': { atomic_cell{regular,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'position': { atomic_cell{-1,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'type': { atomic_cell{frozen<list<frozen<tuple<bigint, frozen<list<blob>>, tinyint>>>>,ts=-9223372036854775807,expiry=-1,ttl=0} }
>         },
>         {
>             'table_name': cdc_generation_descriptions,
>             'column_name': expired,
>             'clustering_order': { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'column_name_bytes': { atomic_cell{65787069726564,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'kind': { atomic_cell{regular,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'position': { atomic_cell{-1,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'type': { atomic_cell{timestamp,ts=-9223372036854775807,expiry=-1,ttl=0} }
>         },
>         {
>             'table_name': cdc_generation_descriptions,
>             'column_name': time,
>             'clustering_order': { atomic_cell{NONE,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'column_name_bytes': { atomic_cell{74696d65,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'kind': { atomic_cell{partition_key,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'position': { atomic_cell{0,ts=-9223372036854775807,expiry=-1,ttl=0} }
>             'type': { atomic_cell{timestamp,ts=-9223372036854775807,expiry=-1,ttl=0} }
>         },
>     ]
> }
> }

File: mutation_partition.cc (hunk 1 above)
Comment (avelanarius) [reply mutation_partition.cc]: Fixed in amended commit (static row before clustering rows).

File: mutation_partition.cc (hunk 2 above)
Comment (avelanarius) [reply mutation_partition.cc:1074]: As this functionality is in two different source files, I have not found an elegant way to refactor it (without adding too much new complexity). Another issue is that the indentation/formatting is different between those instances.

File: mutation.cc [hunk 3]
```diff
@@ -278,7 +278,20 @@ future<mutation_opt> read_mutation_from_flat_mutation_reader(flat_mutation_reade
...
+    auto type_iterator = dk._key.get_compound_type(s)->types().begin();
+    auto column_iterator = s.partition_key_columns().begin();
+
+    for (auto&& e : dk._key.components(s)) {
+        os << "'" << column_iterator->name_as_text() << "': " << (*type_iterator)->to_string(to_bytes(e)) << ", ";
+        ++type_iterator;
+        ++column_iterator;
+    }
+
+    fmt_print(os, "token: {}}}, ", dk._token);
```
Comment (tgrabiec) [inline mutation.cc:294]: There's `dk._key.with_schema(s)` which returns a pretty-printable partition_key::with_schema_wrapper. Could we reuse it?

**Review by tgrabiec — CHANGES_REQUESTED** [review]

File: mutation_partition.cc [hunk 4]
```diff
@@ -1028,19 +1028,63 @@ operator<<(std::ostream& os, const rows_entry::printer& p) {
...
+        os << indent << "},\n";
+    }
+
+    os << indent << "rows: [\n";
+
+    for (const auto& re : mp.clustered_rows()) {
+        os << indent << indent << "{\n";
+
+        const auto& row = re.row();
+        if (!row.marker().is_missing()) {
```
[the remark this answers is not in the document]
Comment (avelanarius) [reply mutation_partition.cc:1061]: Added those fields back in amended commit.
(to see example how it looks currently, please refer to the updated Github PR description: https://github.com/scylladb/scylla/pull/7159#issue-477896999)

File: mutation_partition.cc [hunk 5]
```diff
@@ -1028,19 +1028,63 @@ operator<<(std::ostream& os, const rows_entry::printer& p) {
...
+            os << indent << indent << indent << "marker: " << row.marker() << ",\n";
+        }
+
+        position_in_partition pip(re.position());
+        if (pip.get_clustering_key_prefix()) {
```
[the remark this answers is not in the document]
Comment (avelanarius) [reply mutation_partition.cc:1066]: Added back printing of `_bound_weight` in amended commit. 
(to see example how it looks currently, please refer to the updated Github PR description: https://github.com/scylladb/scylla/pull/7159#issue-477896999)

**Comment by avelanarius** [comment]
@tgrabiec Regarding your comments about not using already available "printers":

1. When using other printers, they introduce a lot of additional nesting characters `{`, `}` making it less pleasant to read. Moreover, they print unnecessary names (`rows_entry`, `deletable_rows`) and no indentation.
2. They print `clustering_key`, `partition_key` in format `field1:field2:...` not desired `'column_name1': field1, 'column_name2': field2, ...`. I was not comfortable changing how clustering, partition keys are printed, because for some log messages it might be better to omit the column names (and other code expected them to print in "old" format).

**Review by tgrabiec — APPROVED** [review]

**Comment by haaawk** [comment]
@tgrabiec Could you merge, please?

**Comment by tgrabiec** [comment]
Merged in https://github.com/scylladb/scylla/commit/561753fe7192645b7cf33f53c2bc7c2cc37b6754

**Pull request closed without merging**